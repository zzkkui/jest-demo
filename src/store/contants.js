export const CHANGE_INPUT_VALUE = 'changeInputValue'

export const CHANGE_ITEMS = 'changeItems'