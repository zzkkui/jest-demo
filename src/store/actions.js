import axios from 'axios'
import { CHANGE_INPUT_VALUE, CHANGE_ITEMS } from './contants'

const actionsBuilder = (type) => (value) => ({
  type,
  value
})

export const changeInputValue = actionsBuilder(CHANGE_INPUT_VALUE)

export const changeItems = actionsBuilder(CHANGE_ITEMS)

// 这里引入了 redux-thunk 所以 action 可以返回一个函数
export const getItems = () => {
  return (dispatch) => {
    return axios.get('/undolist.json').then(res => {
      // const actions = changeItems(res.data)
      // dispatch(actions)
      dispatch({
        type: CHANGE_ITEMS,
        value: res.data
      })
    }).catch(e => {
      // console.log(e)
      dispatch({
        type: CHANGE_ITEMS,
        value: []
      })
    })
  }
}