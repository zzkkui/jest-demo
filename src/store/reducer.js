import { CHANGE_INPUT_VALUE, CHANGE_ITEMS } from './contants'

const initState = {
  inputValue: '',
  undoItems: []
}

export default (state = initState, action) => {
  const { type, value } = action
  // const newState = {...state}
  switch (type) {
    case CHANGE_INPUT_VALUE:
      return { ...state, inputValue: value }
    case CHANGE_ITEMS:
      return { ...state, undoItems: value }
    default:
      return state
  }

}