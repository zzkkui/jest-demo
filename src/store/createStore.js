import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import reducer, { initState } from './reducer'
import * as actions from './actions'

const composeEnhancers =
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) : compose;

const enhancer = composeEnhancers(
  applyMiddleware(thunk)
);

const reducers = combineReducers({ todo: reducer })

const store = createStore(reducers, enhancer)

export { actions, reducers, enhancer }

export default store