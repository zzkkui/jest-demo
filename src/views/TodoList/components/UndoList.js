import React, { Component } from 'react'
import { connect } from 'react-redux'

import { actions } from '../../../store/createStore'

class UndoList extends Component {

  inputRef = React.createRef();

  onClickDelete = (e, index) => {
    e.stopPropagation()
    if (!this.props.undoItems[index].isChecked) {
      const newUndoItems = [...this.props.undoItems]
      newUndoItems.splice(index, 1)
      this.props.changeItems(newUndoItems)
    }
  }

  changeStatus = async (index) => {
    const { undoItems } = this.props;
    if (!undoItems[index].isFocus) {
      const newUndoItems = [...undoItems]
      newUndoItems[index].isFocus = !newUndoItems[index].isFocus
      await this.props.changeItems(newUndoItems)
      this.inputRef.current && this.inputRef.current.focus()
    }
  }

  onBlurInputItem = (index) => {
    const newUndoItems = [...this.props.undoItems]
    newUndoItems[index].isFocus = !newUndoItems[index].isFocus
    this.props.changeItems(newUndoItems)
  }

  changeInputItem = (e, i) => {
    const newUndoItems = [...this.props.undoItems]
    newUndoItems[i] = {
      ...newUndoItems[i],
      value: e.target.value
    }
    this.props.changeItems(newUndoItems)
  }

  changeCheckout = (e, i) => {
    e.stopPropagation()
    const newUndoItems = [...this.props.undoItems]
    newUndoItems[i].isChecked = !newUndoItems[i].isChecked
    this.props.changeItems(newUndoItems)
  }

  render() {
    const { undoItems } = this.props;
    return (
      <div className="undo-list">
        <div className="undo-list-title">
          正在进行
          <div className="undo-list-count" data-test="count">{undoItems.length}</div>
        </div>
        <ul className="undo-list-ul">
          {
            undoItems.map((n, i) =>
              <li className={`${n.isChecked ? "undo-item-checked" : ''} undo-list-item`} data-test="list-item" key={i} onClick={() => this.changeStatus(i)}>
                <input className="undo-list-checkbox" type="checkbox" data-test="check-item" value={n.isChecked} onClick={(e) => this.changeCheckout(e, i)} />
                {
                  n.isFocus ?
                    <input className="undo-list-input" data-test="list-input" ref={this.inputRef} value={n.value} onBlur={() => this.onBlurInputItem(i)} onChange={(e) => this.changeInputItem(e, i)} /> :
                    <span>{n.value}</span>
                }
                <div className="undo-list-delete" data-test="delete-item" onClick={(e) => this.onClickDelete(e, i)}> - </div>
              </li>)
          }
        </ul>
      </div>
    )
  }
}

export default connect(
  ({ todo }) => ({ undoItems: todo.undoItems }),
  { changeItems: actions.changeItems }
)(UndoList)
