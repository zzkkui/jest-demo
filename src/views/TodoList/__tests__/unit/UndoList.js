
import React from 'react';
import { shallow } from 'enzyme';
import UndoList from '../../components/UndoList';
import { findTestWrapper } from '../../../../utils/testUtils'
import store from '../../../../store/createStore'

const listData = [{
  value: 'jest',
  isFocus: true,
  isChecked: false
}, {
  value: 'TDD',
  isFocus: false,
  isChecked: true
}, {
  value: 'react',
  isFocus: false,
  isChecked: false
}]

describe('UndoList 组件', () => {

  it('样式渲染正常', () => {
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData
    })
    expect(warpper).toMatchSnapshot()
  })

  it('传入的 list 为空数组', () => {
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: []
    })
    const countEle = findTestWrapper(warpper, 'count')
    const listItemEle = findTestWrapper(warpper, 'list-item')
    expect(countEle.text()).toBe("0")
    expect(listItemEle.length).toBe(0)
  })

  it('传入的 list 不为空数组，展示数据，有删除按钮，有 checkbox', () => {
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData
    })
    const countEle = findTestWrapper(warpper, 'count')
    const listItemEle = findTestWrapper(warpper, 'list-item')
    const deleteItems = findTestWrapper(warpper, 'delete-item')
    const checkItems = findTestWrapper(warpper, 'check-item')
    expect(countEle.text()).toBe("3")
    expect(listItemEle.length).toBe(3)
    expect(deleteItems.length).toBe(3)
    expect(checkItems.length).toBe(3)
  })

  it('未选中的项，点击删除按钮，会调用传入的 changeItems 方法', () => {
    const fn = jest.fn()
    const index = 1
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      changeItems: fn,
      undoItems: listData
    })
    const deleteItems = findTestWrapper(warpper, 'delete-item')
    deleteItems.at(index).simulate('click', {
      // 阻止事件冒泡
      stopPropagation() { }
    })
    if (!listData[index].isChecked) {
      expect(fn).toHaveBeenLastCalledWith(listData[index])
    }
  })

  it('已经选中的项，点击删除按钮，不会调用传入的 deleteUndoItem 方法', () => {
    const fn = jest.fn()
    const index = 1
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      changeItems: fn,
      undoItems: listData
    })
    const deleteItems = findTestWrapper(warpper, 'delete-item')
    deleteItems.at(index).simulate('click', {
      // 阻止事件冒泡
      stopPropagation() { }
    })
    if (listData[index].isChecked) {
      expect(fn).not.toHaveBeenCalled()
    }
  })

  it('当 isFocus 为 true 这一项被点击时，不触发执行 changeItems 方法', () => {
    const fn = jest.fn()
    const index = 0
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      changeItems: fn,
      undoItems: listData
    })
    const listItemEle = findTestWrapper(warpper, 'list-item')
    listItemEle.at(index).simulate('click')
    // listItemEle.at(index).simulate('focus')
    // 这里只有 isFocus 为 true 的项才会调用 changeItems
    if (listData[index].isFocus) {
      expect(fn).not.toHaveBeenCalled()
    }
  })

  it('当 isFocus 为 false 这一项被点击时，会触发执行 changeItems 方法', () => {
    const fn = jest.fn()
    const index = 1
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      changeItems: fn,
      undoItems: listData
    })
    const listItemEle = findTestWrapper(warpper, 'list-item')
    listItemEle.at(index).simulate('click', {
      stopPropagation() { }
    })
    // listItemEle.at(index).simulate('focus')
    // 这里只有 isFocus 为 true 的项才会调用 changeItems
    if (!listData[index].isFocus) {
      expect(fn).toHaveBeenLastCalledWith([
        listData[0],
        {
          value: 'TDD',
          isFocus: true,
          isChecked: true
        },
        listData[2]])
    }
  })

  it('当某一项 isFocus 为true，展示 input 框', () => {
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData
    })
    const inputItemEle = findTestWrapper(warpper, 'list-input')
    expect(inputItemEle.length).toBe(1)
  })

  it('当某一项输入框失去焦点，触发 changeItems 方法', () => {
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const fn = jest.fn()
    const index = 0
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData,
      changeItems: fn
    })
    const inputItemEle = findTestWrapper(warpper, 'list-input')
    inputItemEle.at(index).simulate('blur')
    expect(fn).toHaveBeenLastCalledWith([
      {
        value: 'jest',
        isFocus: false,
        isChecked: false
      },
      listData[1],
      listData[2]])
  })

  it('当某一项 input change，触发执行传入的 changeItems 方法', () => {
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const fn = jest.fn()
    const index = 0
    const userInput = 'jest-change'
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData,
      changeItems: fn
    })
    const inputItemEle = findTestWrapper(warpper, 'list-input')
    inputItemEle.at(index).simulate('change', {
      target: {
        value: userInput
      }
    })
    expect(fn).toHaveBeenLastCalledWith([
      {
        value: userInput,
        isFocus: true,
        isChecked: false
      },
      listData[1],
      listData[2]])
  })

  it('当某一项 checkout 选中，触发 changeItems 方法', () => {
    const listData = [{
      value: 'jest',
      isFocus: true,
      isChecked: false
    }, {
      value: 'TDD',
      isFocus: false,
      isChecked: true
    }, {
      value: 'react',
      isFocus: false,
      isChecked: false
    }]
    const fn = jest.fn()
    const index = 0
    const warpper = shallow(<UndoList store={store} />).find('UndoList').dive()
    warpper.setProps({
      undoItems: listData,
      changeItems: fn
    })
    const checkItemEle = findTestWrapper(warpper, 'check-item')
    checkItemEle.at(index).simulate('click', {
      stopPropagation() { }
    })
    expect(fn).toHaveBeenLastCalledWith([
      {
        value: 'jest',
        isFocus: true,
        isChecked: true
      },
      listData[1],
      listData[2]])
  })
});


