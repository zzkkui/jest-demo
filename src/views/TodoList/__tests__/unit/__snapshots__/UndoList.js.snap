// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList 组件 样式渲染正常 1`] = `
<div
  className="undo-list"
>
  <div
    className="undo-list-title"
  >
    正在进行
    <div
      className="undo-list-count"
      data-test="count"
    >
      3
    </div>
  </div>
  <ul
    className="undo-list-ul"
  >
    <li
      className=" undo-list-item"
      data-test="undo-list-item"
      key="0"
      onClick={[Function]}
    >
      <input
        className="undo-list-checkbox"
        data-test="check-item"
        onClick={[Function]}
        type="checkbox"
        value={false}
      />
      <input
        className="undo-list-input"
        data-test="list-input"
        onBlur={[Function]}
        onChange={[Function]}
        value="jest"
      />
      <div
        className="undo-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
         - 
      </div>
    </li>
    <li
      className="undo-item-checked undo-list-item"
      data-test="undo-list-item"
      key="1"
      onClick={[Function]}
    >
      <input
        className="undo-list-checkbox"
        data-test="check-item"
        onClick={[Function]}
        type="checkbox"
        value={true}
      />
      <span>
        TDD
      </span>
      <div
        className="undo-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
         - 
      </div>
    </li>
    <li
      className=" undo-list-item"
      data-test="undo-list-item"
      key="2"
      onClick={[Function]}
    >
      <input
        className="undo-list-checkbox"
        data-test="check-item"
        onClick={[Function]}
        type="checkbox"
        value={false}
      />
      <span>
        react
      </span>
      <div
        className="undo-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
         - 
      </div>
    </li>
  </ul>
</div>
`;
